native/cocos/audio/apple/AudioCache.mm

```diff
@@ -23,284 +23,229 @@ of this software and associated engine source code (the "Software"), a limited,
-    _state = State::LOADING;
-
-    AudioDecoder decoder;
+// TODO: If is streaming audio, return partial data? or whole data.
+ccstd::vector<uint8_t> AudioCache::getPCMBuffer(uint32_t channelID){
+    ccstd::vector<uint8_t> ret;
     do {
-        if (!decoder.open(_fileFullPath.c_str()))
+        if (loadState != State::LOADED) {
+            NSLog(@"AudioCache is not ready");
             break;
-
-        const uint32_t originalTotalFrames = decoder.getTotalFrames();
-        const uint32_t bytesPerFrame = decoder.getBytesPerFrame();
-        const uint32_t sampleRate = decoder.getSampleRate();
-        const uint32_t channelCount = decoder.getChannelCount();
-
-        uint32_t totalFrames = originalTotalFrames;
-        uint32_t dataSize = totalFrames * bytesPerFrame;
-        uint32_t remainingFrames = totalFrames;
-        uint32_t adjustFrames = 0;
-
-        _format = channelCount > 1 ? AL_FORMAT_STEREO16 : AL_FORMAT_MONO16;
-        _sampleRate = (ALsizei)sampleRate;
-        _duration = 1.0f * totalFrames / sampleRate;
-        _totalFrames = totalFrames;
-
-        if (dataSize <= PCMDATA_CACHEMAXSIZE) {
-            uint32_t framesRead = 0;
-            const uint32_t framesToReadOnce = std::min(totalFrames, static_cast<uint32_t>(sampleRate * QUEUEBUFFER_TIME_STEP * QUEUEBUFFER_NUM));
-
-            BREAK_IF_ERR_LOG(!decoder.seek(totalFrames), "AudioDecoder::seek(%u) error", totalFrames);
-
-            char *tmpBuf = (char *)malloc(framesToReadOnce * bytesPerFrame);
-            ccstd::vector<char> adjustFrameBuf;
-            adjustFrameBuf.reserve(framesToReadOnce * bytesPerFrame);
-
-            // Adjust total frames by setting position to the end of frames and try to read more data.
-            // This is a workaround for https://github.com/cocos2d/cocos2d-x/issues/16938
-            do {
-                framesRead = decoder.read(framesToReadOnce, tmpBuf);
-                if (framesRead > 0) {
-                    adjustFrames += framesRead;
-                    adjustFrameBuf.insert(adjustFrameBuf.end(), tmpBuf, tmpBuf + framesRead * bytesPerFrame);
+        }
+        if (channelID > _pcmHeader.channelCount) {
+            NSLog(@"ChannelID is bigger than channel count");
+            break;
+        }
+        
+        AVAudioPCMBuffer* tmpBuffer = [[AVAudioPCMBuffer alloc] initWithPCMFormat:_descriptor.audioFile.processingFormat frameCapacity:_descriptor.audioFile.length];
+        // Read entire buffer into _descriptor buffer.
+        [_descriptor.audioFile readIntoBuffer:tmpBuffer error:nil];
+        const uint32_t frameCount = _pcmHeader.totalFrames;
+        ret.resize(_pcmHeader.bytesPerFrame * _pcmHeader.totalFrames / _pcmHeader.channelCount);
+        uint8_t *buffer = ret.data();
+        
+        uint32_t bitLength = _pcmHeader.bytesPerFrame / _pcmHeader.channelCount * 8;
+        switch (_pcmHeader.dataFormat) {
+            // TODO: if float 32 and float 63 both convert to float 32 by default?
```

Comment:
> The floatChannelData property returns pointers to the buffer’s audio samples if the buffer’s format is 32-bit float. It returns nil if it’s another format.

https://developer.apple.com/documentation/avfaudio/avaudiopcmbuffer/1386212-floatchanneldata?language=objc
According to the link floatChannelData may return null in float64 mode.